Calendar editor rewrite 5437

ganthern opened the pull request:
at some point, 
fix #5437

ganthern commented on src/api/common/TutanotaConstants.ts:
```diff
@@ -451,7 +451,8 @@ export const enum EndType {
 export const defaultCalendarColor = "2196f3"
 
-export const enum AlarmInterval {
+// keep non-const for iteration/filtering
```
like, duplicate the values into an object and filter on that? copying at runtime is not really possible with a const enum right?

wrdhub replied:
Okay, I see. I didn't realize how const enums were compiled

charlag commented on src/calendar/view/eventeditor/AttendeeListEditor.ts:
```diff
@@ -0,0 +1,277 @@
+		: m(IconButton, {
+				title: "sendMail_alt",
+				click: async () =>
+					(await import("../../../contacts/view/ContactView.js")).writeMail(
```
oof this seems like something that should be somehow injected idk. maybe it's fine for now but is a bit dirty

ganthern replied:
i felt it was too deeply nested to pass it all the way down

charlag replied:
just passing it from the editor dialog would be okay, I don't think deeply nested is a good counter argument but also it doesn't matter, if it's too much pain then just leave it

charlag commented on src/calendar/view/eventeditor/CalendarEventEditDialog.ts:
```diff
@@ -0,0 +1,235 @@
+		}
+
+		if (
+			model.eventType === EventType.OWN &&
```
can this login not be here but elsewhere?

ganthern replied:
I'm not sure I understand, there's no login here

charlag replied:
I probably meant logic oops

charlag commented on src/calendar/view/eventeditor/CalendarEventEditView.ts:
```diff
@@ -0,0 +1,280 @@
+				icon: BootIcons.Expand,
+				disabled: model.eventType !== EventType.OWN && model.eventType !== EventType.SHARED_RW,
+				helpLabel: () => this.renderCalendarColor(model.editModels.whoModel.selectedCalendar, vnode.attrs.groupColors),
+			} as DropDownSelectorAttrs<CalendarInfo>),
```
others are `satisfies` but this one is a cast?

charlag commented on src/calendar/view/eventeditor/RepeatRuleEditor.ts:
```diff
@@ -0,0 +1,154 @@
+}
+
+export class RepeatRuleEditor implements Component<RepeatRuleEditorAttrs> {
+	private readonly startOfTheWeekOffset: number
```
why do we have to keep it around?

ganthern replied:
we don't, fixed

ganthern commented on src/calendar/date/CalendarInvites.ts:
```diff
@@ -1,18 +1,26 @@
+import { CalendarEventModel, EventType } from "./eventeditor/CalendarEventModel.js"
 
-function getParsedEvent(fileData: DataFile):
+export type Guest = Pick<CalendarEventAttendee, "status"> & Recipient & { status: CalendarAttendeeStatus }
```
removed the pick

ganthern commented on src/calendar/date/CalendarUtils.ts:
```diff
@@ -562,31 +554,40 @@ export function getStartOfTheWeekOffsetForUser(userSettingsGroupRoot: UserSettin
+	}
+}
+
+/** in: some {@param date} object that encodes some calendar date in {@param zone} (like the 1st of May 2023)
```
improved wording

ganthern commented on src/calendar/date/CalendarUtils.ts:
```diff
@@ -1107,23 +1117,30 @@ export function calendarAttendeeStatusSymbol(status: CalendarAttendeeStatus): st
 }
 
-export function incrementSequence(sequence: string, isOwnEvent: boolean): string {
+export const iconForAttendeeStatus: Record<CalendarAttendeeStatus, AllIcons> = Object.freeze({
```
we could try identifying the areas of responsibility it covers and splitting it up, it's fairly large

ganthern commented on src/calendar/date/CalendarUtils.ts:
````diff
@@ -1137,16 +1154,19 @@ export function findPrivateCalendar(calendarInfo: Map<Id, CalendarInfo>): Calend
  * ```
+ *
+ * @param description description to clean up
+ * @param sanitizer optional sanitizer to apply after preparing the description
````
agreed

ganthern commented on src/calendar/date/CalendarUtils.ts:
```diff
@@ -1206,7 +1228,7 @@ export function isEventBetweenDays(event: CalendarEvent, firstDay: Date, lastDay
 }
 
-export function createRepeatRuleFrequencyValues(): SelectorItemList<RepeatPeriod | null> {
+export const createRepeatRuleFrequencyValues = lazyMemoized((): SelectorItemList<RepeatPeriod | null> => {
```
did not think of that at all. removed the memoization from the arrays

ganthern commented on src/calendar/date/eventeditor/CalendarEventAlarmModel.ts:
```diff
@@ -0,0 +1,83 @@
+ * edit the alarms set on a calendar event.
+ */
+export class CalendarEventAlarmModel {
+	private _alarms: Set<AlarmInterval> = new Set()
```
I get that and I'd argue that we should still try to dedupe them even with custom values or distinction between different kinds of alarm. We'll probably have some UI in front to enforce syntax and then it should be fairly easy.

but that discussion is for another day :)

ganthern commented on src/calendar/date/eventeditor/CalendarEventAlarmModel.ts:
```diff
@@ -0,0 +1,83 @@
+	 * get the alarm triggers that are currently set on the event.
+	 */
+	get alarms(): Array<AlarmInterval> {
+		return Array.from(this._alarms.values())
```
surprisingly, it's unused because only splitTriggers is used

ganthern commented on src/calendar/date/eventeditor/CalendarEventAlarmModel.ts:
```diff
@@ -0,0 +1,83 @@
+}
+
+export function newAlarm(trigger: AlarmInterval): AlarmInfo {
+	return createCalendarAlarm(generateEventElementId(Date.now()), trigger)
```
it's injected now

ganthern commented on src/calendar/date/eventeditor/CalendarEventAlarmModel.ts:
```diff
@@ -0,0 +1,83 @@
+}
+
+export function createCalendarAlarm(identifier: string, trigger: string): AlarmInfo {
+	const calendarAlarmInfo = createAlarmInfo()
```
yep

ganthern commented on src/calendar/date/eventeditor/CalendarEventWhoModel.ts:
```diff
@@ -0,0 +1,637 @@
+		const ownAddresses = this.ownMailAddresses.map((a) => cleanMailAddress(a.address))
+
+		// convert the list of attendees into a map for easier use.
+		for (const attendee of initialValues.attendees?.slice().map((a) =>
```
i might have been asleep

ganthern commented on src/calendar/date/eventeditor/CalendarEventWhoModel.ts:
```diff
@@ -0,0 +1,637 @@
+					// the organizer added themselves, but did not specify if they're participating
+					status: CalendarAttendeeStatus.NEEDS_ACTION,
+				})
+			this.initialAttendees.delete(this._organizer!.address.address)
```
indeed

ganthern commented on src/calendar/date/eventeditor/CalendarEventWhoModel.ts:
```diff
@@ -0,0 +1,637 @@
+			Array.from(this._attendees.values()).some((a) => a.status !== CalendarAttendeeStatus.ADDED)
+		) {
+			// this is technically an invalid state that should not happen.
+			console.log("got an event with attendees and an organizer that's not the owner of the calendar, replacing organizer.")
```
it's invalid now, but before I think we always set the organizer to the person creating the event and replaced it when the actual calendar owner edited the event. clarified the comment.

ganthern commented on src/calendar/date/eventeditor/CalendarEventWhoModel.ts:
```diff
@@ -0,0 +1,637 @@
+			!this.isNew &&
+			// if the event is not new, but the attendee list did not have any attendees that were already notified,
+			// there are no attendees that are not either us or the organizer
+			Array.from(this.initialAttendees.values()).filter((a) => a.status !== CalendarAttendeeStatus.ADDED).length > 0
```
but I like my hammer

ganthern commented on src/calendar/date/eventeditor/CalendarEventWhoModel.ts:
```diff
@@ -0,0 +1,637 @@
+			return [this._organizer.address]
+		} else {
+			// something is wrong.
+			return []
```
it's a ProgrammingError now

ganthern commented on src/calendar/date/eventeditor/CalendarEventWhoModel.ts:
```diff
@@ -0,0 +1,637 @@
+	 */
+	private addOwnAttendee(address: EncryptedMailAddress): void {
+		if (this.hasOtherAttendees()) {
+			console.log("can't change organizer if there are other invitees already")
```
yes, if you're adding yourself and get to this place, you're the organizer and are adding another of your addresses.

ganthern commented on src/calendar/date/eventeditor/CalendarEventWhoModel.ts:
```diff
@@ -0,0 +1,637 @@
+ * @param after the map that's representing the current contents.
+ * @returns arrays containing the kept, added, and deleted values.
+ */
+export function trisectingDiff<T>(before: ReadonlyMap<unknown, T>, after: ReadonlyMap<unknown, T>): { kept: Array<T>; added: Array<T>; deleted: Array<T> } {
```
I am not worthy

ganthern commented on src/calendar/date/eventeditor/CalendarEventWhoModel.ts:
```diff
@@ -0,0 +1,637 @@
+
+	return {
+		allAttendees,
+		organizerToPublish: organizer.address,
```
it might be passed back as null, or am I misunderstanding the problem?

charlag replied:
oh I see now, sorry!

ganthern commented on src/calendar/view/CalendarView.ts:
```diff
@@ -806,6 +791,35 @@ export class CalendarView extends BaseTopLevelView implements TopLevelView<Calen
+ * @param calendarEvent
+ * @param entityClient
+ */
+export async function resolveCalendarEventProgenitor(calendarEvent: CalendarEvent, entityClient: EntityClient = locator.entityClient): Promise<CalendarEvent> {
```
fixed

ganthern commented on src/calendar/view/eventeditor/AttendeeListEditor.ts:
```diff
@@ -0,0 +1,277 @@
+
+export type AttendeeListEditorAttrs = {
+	/** the event that is currently being edited */
+	model: CalendarEventModel
```
only for the shouldShowSendInviteNotAvailable call and the shouldSendUpdates checkbox. the former could be injected in the attrs I guess and the latter could be moved to the CalendarEventWhoModel since it's preparing the SendMailModels anyway.

charlag replied:
I think it's okay to keep it

ganthern commented on src/calendar/view/eventeditor/CalendarEventEditView.ts:
```diff
@@ -0,0 +1,280 @@
+	private renderRemindersEditor(vnode: Vnode<CalendarEventEditViewAttrs>): Children {
+		// we can set reminders only if we're able to edit the event on the server because we have to add them to the entity.
+		if (
+			vnode.attrs.model.eventType !== EventType.OWN &&
```
suggestion: put a canEditReminders on the model?

charlag replied:
sounds good

ganthern commented on src/gui/base/buttons/BannerButton.ts:
```diff
@@ -0,0 +1,58 @@
+	}
+}
+
+export function renderReplyButtons(participation: NonNullable<EventPreviewViewAttrs["participation"]>) {
```
or use segmentControl maybe

charlag replied:
segmented control doesn't usually have "nothing selected" state, it's not it

ganthern replied:
I guess you cold say that, but I doubt that's a general guideline. To me it would make sense to use a segment control for a choice that can be made once and only changed, but not completely taken back later

I made a pureComponent out of it and moved it to the EventPreviewView, which took some doing since mail-view and calendar-view can't import each other and putting the import all the way at the top of the mailview hierarchy just feels completely wrong.

we may need an asyncComponent utility in the future to make things like this easier, maybe with a placeholder feature so there's no UI shift even if the import takes a bit.

ganthern commented on src/gui/base/InfoBanner.ts:
```diff
@@ -0,0 +1,116 @@
+	helpLink?: InfoLink | null
+	buttons: ReadonlyArray<ButtonAttrs | null>
+	type?: BannerType
+	logins: LoginController
```
seemed weird to me as well, the helpLink is already nullable so we can make that decision somewhere else.

ganthern commented:
as for the hashUid conundrum, my current state is this:
* i do not ever have an event that does not have a UID, but much less do I want an event that has a UID but no hashedUID
* but hashedUid is also not really relevant to the client   
-> I need to add the hashedUid during construction of events or before they leave the client with a server call.

* enforcing presence of hashedUid in every public method of CalendarFacade is not enough because people are calling calendarEventResource directly.
* for the same reason it's not possible to have the hashedUid as a separate argument to each server call that takes an event with Uid, they're resources not services
* so we opt for ensuring the presence at construction (assignEventIdentity, in the parser/importer)

-> looks like at least the uid/hashedUid generator needs to move to the worker

charlag commented:
for now the easiest thing to do with `hashedUid` is to move it back to calendarFacade and just update those place which create events to be `async`. I think we all agree that passing around current `CalendarEvent` entity is just a very leaky and error-prone abstraction and we should have some wrapper which also knows which kind of "handle" it is (e.g. the original event, occurrence etc), has assessors for things like all-day marker etc and can also enforce `hashedUid` invariant.

If we only need `hashedUid` for the server we could also just say that we always setup events via `CalendarFacade` (and maybe through `CalendarModel`, how it was) and just let it take care of it.

ganthern commented on src/calendar/date/eventeditor/CalendarEventAlarmModel.ts:
```diff
@@ -11,9 +11,9 @@ export type CalendarEventAlarmModelResult = {
+	private readonly _alarms: Set<AlarmInterval> = new Set()
 
-	constructor(alarms: Array<AlarmInterval> = [], private readonly uiUpdateCallback: () => void = noOp) {
+	constructor(alarms: Array<AlarmInterval> = [], private readonly now: () => number, private readonly uiUpdateCallback: () => void = noOp) {
```
yes, should use those, I amended the commit

ganthern commented:
> for now the easiest thing to do with `hashedUid` is to move it back to calendarFacade and just update those place which create events to be `async`. I think we all agree that passing around current `CalendarEvent` entity is just a very leaky and error-prone abstraction and we should have some wrapper which also knows which kind of "handle" it is (e.g. the original event, occurrence etc), has assessors for things like all-day marker etc and can also enforce `hashedUid` invariant.

agreed, naked CalendarEvents shouldn't really be used/inspected on their own at some point.

> If we only need `hashedUid` for the server we could also just say that we always setup events via `CalendarFacade` (and maybe through `CalendarModel`, how it was) and just let it take care of it.

I'd actually argue that storing it on the entity at all seems weird, hence my indecisiveness what to do with it.

charlag commented on src/calendar/export/CalendarImporter.ts:
```diff
@@ -92,6 +72,63 @@ export function serializeCalendar(
+//
+
+/** importer internals exported for testing */
+export async function parseCalendarStringData(value: string, zone: string): Promise<ParsedCalendarData> {
```
did this just become async? i see a lot of changed lines in this file but not lots of changes

ganthern replied:
yes, I was confused about the public interface of this thing and moved stuff around. must have missed that I left the now pointless async in. amended the commit to fix it.

ganthern commented on src/calendar/view/eventeditor/AttendeeListEditor.ts:
```diff
@@ -0,0 +1,277 @@
+					.filter((a) => a.type === RecipientType.EXTERNAL)
+					.map((guest) => {
+						const { address } = guest
+						if (!whoModel.getPresharedPassword(address)) whoModel.setPresharedPassword(address, "")
```
it's already happening there, that condition is never true. oops!

ganthern commented on src/calendar/date/eventeditor/CalendarEventModel.ts:
```diff
@@ -0,0 +1,690 @@
+		now.invitedConfidentially !== previous.invitedConfidentially ||
+		now.uid !== previous.uid ||
+		!areRepeatRulesEqual(now.repeatRule, previous?.repeatRule ?? null) ||
+		!arrayEqualsWithPredicate(
```
it does, mainly because the owner/organizer might change their status and save it.

charlag commented on src/calendar/date/eventeditor/CalendarEventWhenModel.ts:
```diff
@@ -0,0 +1,534 @@
+	 * will also modify the excluded dates if there are any to still exclude the
+	 * same occurrence dates.
+	 */
+	set isAllDay(value: boolean) {
```
I think it's  marginally better but I don't want to insist on it

charlag commented:
Desktop app just does not run, for some reason `MainLocator` is included in the destkop, probably because of some calendar utils

charlag commented:
![Screenshot from 2023-06-08 10-22-49](https://github.com/tutao/tutanota/assets/3099142/c762cbff-fd9e-4e59-8be6-d239f4c78861)

charlag commented:
I've saved an event in personal calendar,  then I've added a guest. The email was never sent, also not when I update the event

charlag commented:
Cancellation *was* sent tho

ganthern commented:
> I've saved an event in personal calendar, then I've added a guest. The email was never sent, also not when I update the event

that should work now.

> ![Screenshot from 2023-06-08 10-22-49](https://user-images.githubusercontent.com/3099142/244312067-c762cbff-fd9e-4e59-8be6-d239f4c78861.png)

fixed
